core/src/toga/sources/tree_source.py

```diff
@@ -1,13 +1,34 @@
 from __future__ import annotations
 
-from typing import Any
+import sys
+from collections.abc import Iterator
+from typing import TYPE_CHECKING, Generic, Iterable, Mapping, Tuple, TypeVar, Union
 
 from .base import Source
 from .list_source import Row, _find_item
 
+T = TypeVar("T")
 
-class Node(Row):
-    def __init__(self, **data):
+if TYPE_CHECKING:
+    if sys.version_info < (3, 10):
+        from typing_extensions import TypeAlias
+    else:
+        from typing import TypeAlias
+
+    NodeDataT: TypeAlias = Union[object, Mapping[str, T], Iterable[T]]
+    TreeSourceDataT: TypeAlias = Union[
+        object,
+        Mapping[NodeDataT[T], "TreeSourceDataT[T]"],
+        Iterable[Tuple[NodeDataT[T], "TreeSourceDataT[T]"]],
+    ]
+else:
+    TreeSourceDataT = None
```

Comment:
As a note, we end up with weird `TYPE_CHECKING` blocks like this if we continue managing `typing_extensions` the way. That is, by including it as a `dev` requirement, we can't freely use `TypeAlias` (or anything else we'll need from it) and as a consequence, we can't assume anything defined as a `TypeAlias` is available elsewhere for importing.

Interestingly, I didn't know I had all these import issues until I actually tried running a real app with Python 3.9. Neither the unit tests or testbed exercise in CI caught it.

The alternative would be to add `typing_extensions` as a full requirement....but I definitely understand the reluctance there.

Just as a further note, the design of `typing_extensions` basically lets you use it exclusively. The devs simply alias the backported classes to the real deal once youre on a supported Python. Furthermore, though, they backport features of classes; so, `TypeVar`, for instance, has features in 3.12 that do not exist in stdlib in 3.10. But if you use the `TypeVar` from `typing_extensions`, you get those features even on 3.10.

Finally, pyupgrade supports `typing_extensions` and will automatically revert its use to `typing` when only appropriate Pythons are supported.

Mostly just wanted to lay this out in one place.